// @sap/cds/lib/auth/index.js

const cds = require ('../index'), { path, local } = cds.utils

const _require = require; require = cds.lazified (module) // eslint-disable-line no-global-assign
module.exports = Object.assign (auth_factory, {
  mocked: require('./basic-auth'),
  basic:  require('./basic-auth'),
  dummy:  require('./dummy-auth'),
  ias:    require('./ias-auth'),
  jwt:    require('./jwt-auth'),
  xsuaa:  require('./jwt-auth'),
})
require = _require // eslint-disable-line no-global-assign


/**
 * Constructs one of the above middlewares as configured
 */
function auth_factory (options) {
  const o = { ...options, ...cds.requires.auth }
  let kind = o.kind || o.strategy
  let middleware = cds.auth[kind]
  if (middleware) {
    cds.log().info ('using auth strategy:', { kind }, '\n')
  } else {
    let impl = o.impl || path.resolve (__dirname, kind)
    try { impl = require.resolve (impl) }  catch {
      throw cds.error `Cannot find auth impl: ${impl}`
    }
    cds.log().info ('using auth strategy:', { kind, impl: local(impl) }, '\n')
    middleware = require(impl)
  }
  if ((typeof middleware === 'function' && middleware.length === 3) || Array.isArray(middleware)) {
    return middleware
  }
  return middleware(o)
}
