// @sap/hdi/lib/result/MakeGroupResultTuple.js
'use strict';
const ResultTuple = require('../ResultTuple');

/**
 * Corresponds to a TT_MAKE_GROUPS row
 */
class MakeGroupResult {    /**
    * Constructs a JavaScript object representing a table row.
    * @param {*} resultSet The result of a sql procedure call.
    */
  constructor(resultSet) {
    this.makeGroup = resultSet.MAKE_GROUP;
    this.path = resultSet.PATH;
  }

  /**
   * Return a string representation of this instance.
   * @returns {string} Properties of this object as a string.
   */
  toString() {
    return `${this.path  }: ${  this.makeGroup}`;
  }
}
/**
 * An object of this class contains the usual ResultTuple attributes like return code, request ID and messages.
 * The results property is an array of {@link MakeGroupResult}s.
 * @extends ResultTuple
 */
class MakeGroupResultTuple extends ResultTuple {
  /**
   * Creates an instance of MakeGroupResultTuple.
   * @memberof MakeGroupResultTuple
   */
  constructor(){
    super();
  }

  /**
   * Fills the result property of the object with the supplied result set.
   * @param {Object[]} resultSet The result set of a procedure call.
   * @returns {undefined} Returns nothing/undefined.
   * @throws Throws an error, if the result set is not an array.
   */
  fillResultFromResultSet(resultSet) {
    if (resultSet instanceof Array) {
      resultSet.forEach((res) => {
        this.results.push(new MakeGroupResult(res));
      });
    } else {
      throw new Error('Result set should be an array.');
    }
  }
}



module.exports = MakeGroupResultTuple;
