// @sap/hdi/lib/result/MetadataContentResultTuple.js
'use strict';

const ResultTuple = require('../ResultTuple');
const MetadataResult = require('./MetadataResult');

/**
 * A ResultSet row corresponding to TT_FILESFOLDERS_METADATA_CONTENT
 */
class MetadataContentResult extends MetadataResult {    /**
    * Constructs a JavaScript object representing a table row.
    * @param {*} resultSet The result of a sql procedure call.
    */
  constructor(resultSet) {
    super(resultSet);
    this.content = resultSet.CONTENT;
  }

  /**
   * Return a string representation of this instance.
   * @returns {string} Properties of this object as a string.
   */
  toString() {
    return `${super.toString()  }Content:\n${  this.content  }\n`;
  }
}
/**
 * An object of this class contains the usual ResultTuple attributes like return code, request ID and messages.
 * The results property is an array of {@link MetadataContentResult}s.
 * @extends ResultTuple
 */
class MetadataContentResultTuple extends ResultTuple {
  /**
   * Creates an instance of MetadataContentResultTuple.
   * @memberof MetadataContentResultTuple
   */
  constructor(){
    super();
  }

  /**
   * Fills the result property of the object with the supplied result set.
   * @param {Object[]} resultSet The result set of a procedure call.
   * @returns {undefined} Returns nothing/undefined.
   * @throws Throws an error, if the result set is not an array.
   */
  fillResultFromResultSet(resultSet) {
    if (resultSet instanceof Array) {
      resultSet.forEach((res) => {
        this.results.push(new MetadataContentResult(res));
      });
    } else {
      throw new Error('Result set should be an array.');
    }
  }
}
module.exports = MetadataContentResultTuple;
