// @sap/hdi/lib/result/MetadataResult.js
'use strict';

/**
 * A ResultSet row corresponding to TT_FILESFOLDERS_METADATA
 */
class MetadataResult {
  /**
   * Constructs a JavaScript object representing a table row.
   * @param {*} resultSet The result of a sql procedure call.
   */
  constructor(resultSet) {
    this.path = resultSet.PATH;
    this.createUserName = resultSet.CREATE_USER_NAME;
    this.createAUserName = resultSet.CREATE_APPUSER_NAME;
    this.createTimestampUTC = resultSet.CREATE_TIMESTAMP_UTC;
    this.modificationUserName = resultSet.MODIFICATION_USER_NAME;
    this.modificationAUserName = resultSet.MODIFICATION_APPUSER_NAME;
    this.modificationTimestampUTC = resultSet.MODIFICATION_TIMESTAMP_UTC;
    this.size = resultSet.SIZE;
    this.sha256 = resultSet.SHA256;
  }

  /**
   * Return a string representation of this instance.
   * @returns {string} Properties of this object as a string.
   */
  toString() {
    return `\nPath:                     ${  this.path  }\n` +
                `createUserName:           ${  this.createUserName  }\n` +
                `createAppUserName:        ${  this.createAUserName  }\n` +
                `createTimestampUTC:       ${  this.createTimestampUTC  }\n` +
                `modificationUserName:     ${  this.modificationUserName  }\n` +
                `modificationAppUserName:  ${  this.modificationAUserName  }\n` +
                `modificationTimestampUTC: ${  this.modificationTimestampUTC  }\n` +
                `Size:                     ${  this.size  }\n` +
                `SHA256:                   ${  this.sha256  }\n`;
  }
}

module.exports = MetadataResult;
