// @sap/hdi/lib/result/StatusResultTuple.js
'use strict';

const ResultTuple = require('../ResultTuple');

/**
 * A ResultSet row corresponding to TT_FILESFOLDERS_STATUS
 */
class StatusResult {
  /**
   * Constructs a JavaScript object representing a table row.
   * @param {*} resultSet The result of a sql procedure call.
   */
  constructor(resultSet) {
    this.path = resultSet.PATH;
    this.status = resultSet.STATUS;
    this.sha256 = resultSet.SHA256;
    this.sha256Deployed = resultSet.SHA256_DEPLOYED;
  }

  /**
   * Return a string representation of this instance.
   * @returns {string} Properties of this object as a string.
   */
  toString() {
    return `\n${  this.path  } (${  this.status  })`;
  }
}
/**
 * An object of this class contains the usual ResultTuple attributes like return code, request ID and messages.
 * The results property is an array of {@link StatusResult}s.
 * @extends ResultTuple
 */
class StatusResultTuple extends ResultTuple {
  /**
   * Creates an instance of StatusResultTuple.
   * @memberof StatusResultTuple
   */
  constructor() {
    super();
  }

  /**
   * Fills the result property of the object with the supplied result set.
   * @param {Object[]} resultSet The result set of a procedure call.
   * @returns {undefined} Returns nothing/undefined.
   * @throws Throws an error, if the result set is not an array.
   */
  fillResultFromResultSet(resultSet) {
    if (resultSet instanceof Array) {
      resultSet.forEach((res) => {
        this.results.push(new StatusResult(res));
      });
    } else {
      throw new Error('Result set should be an array.');
    }
  }
}

module.exports = StatusResultTuple;
