// @sap/hdi/lib/util/Constants.js
'use strict';


/**
 * This class provides the names of the temporary tables used to pass parameters to the API.
 *
 * @class Constants
 */
class Constants{
  /**
   * Creates an instance of Constants.
   * @memberof Constants
   */
  constructor(){
    this.API_PRIVILEGES_TABLE = '#API_PRIVILEGES';
    this.CONFIG_PARAMETERS_TABLE = '#CONFIG_PARAMETERS';
    this.DEPLOY_PATHS_TABLE = '#DEPLOY_PATHS';
    this.LIBRARY_CONFIG_TABLE = '#LIBRARY_CONFIG';
    this.PARAMETERS_TABLE = '#PARAMETERS';
    this.PATHS_CONTENT_TABLE = '#PATHS_CONTENT';
    this.PATH_PARAMETERS_TABLE = '#PATH_PARAMETERS';
    this.PATHS_TABLE = '#PATHS';
    this.SCHEMA_PRIVILEGES_TABLE = '#SCHEMA_PRIVILEGES';
    this.SCHEMA_ROLES_TABLE = '#SCHEMA_ROLES';
    this.UNDEPLOY_PATHS_TABLE = '#UNDEPLOY_PATHS';
    this.OBJECTS_TABLE = '#OBJECTS';

    this.SYS_DI_SCHEMA = '_SYS_DI';

    this.DI_DEFAULT_COMMON_PRIVILEGES = `${this.SYS_DI_SCHEMA  }.T_DEFAULT_COMMON_PRIVILEGES`;
    this.DI_ADMIN_PRIVILEGES_TABLE = `${this.SYS_DI_SCHEMA  }.T_DEFAULT_DI_ADMIN_PRIVILEGES`;
    this.DEFAULT_CONTAINER_ROLE_ADMIN_TABLE = `${this.SYS_DI_SCHEMA  }.T_DEFAULT_CONTAINER_ROLE_ADMIN_PRIVILEGES`;
    this.GROUP_ADMIN_PRIVILEGES_TABLE = `${this.SYS_DI_SCHEMA  }.T_DEFAULT_CONTAINER_GROUP_ADMIN_PRIVILEGES`;
    this.CONTAINER_ADMIN_PRIVILEGES_TABLE = `${this.SYS_DI_SCHEMA  }.T_DEFAULT_CONTAINER_ADMIN_PRIVILEGES`;
    this.CONTAINER_USER_PRIVILEGES_TABLE = `${this.SYS_DI_SCHEMA  }.T_DEFAULT_CONTAINER_USER_PRIVILEGES`;
    this.DEFAULT_LIBRARIES_TABLE = `${this.SYS_DI_SCHEMA  }.T_DEFAULT_LIBRARIES`;

    this.HELPER_TABLE_TYPES = new Map([
      [this.PARAMETERS_TABLE, 'TT_PARAMETERS'],
      [this.CONFIG_PARAMETERS_TABLE, 'TT_PARAMETERS'],
      [this.PATHS_TABLE, 'TT_FILESFOLDERS'],
      [this.DEPLOY_PATHS_TABLE, 'TT_FILESFOLDERS'],
      [this.UNDEPLOY_PATHS_TABLE, 'TT_FILESFOLDERS'],
      [this.PATH_PARAMETERS_TABLE, 'TT_FILESFOLDERS_PARAMETERS'],
      [this.PATHS_CONTENT_TABLE, 'TT_FILESFOLDERS_CONTENT'],
      [this.API_PRIVILEGES_TABLE, 'TT_API_PRIVILEGES'],
      [this.SCHEMA_PRIVILEGES_TABLE, 'TT_SCHEMA_PRIVILEGES'],
      [this.SCHEMA_ROLES_TABLE, 'TT_SCHEMA_ROLES'],
      [this.LIBRARY_CONFIG_TABLE, 'TT_LIBRARY_CONFIGURATION'],
      [this.OBJECTS_TABLE, 'TT_OBJECTS']
    ]);
  }
}
// Only export an instance of Constants -> Singleton.
module.exports = new Constants();
