// @sap-cloud-sdk/connectivity/dist/scp-cf/destination/destination-from-vcap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.searchServiceBindingForDestination = exports.destinationForServiceBinding = void 0;
const util_1 = require("@sap-cloud-sdk/util");
const jwt_1 = require("../jwt");
const environment_accessor_1 = require("../environment-accessor");
const http_proxy_util_1 = require("./http-proxy-util");
const destination_service_types_1 = require("./destination-service-types");
const destination_cache_1 = require("./destination-cache");
const service_binding_to_destination_1 = require("./service-binding-to-destination");
const logger = (0, util_1.createLogger)({
    package: 'connectivity',
    messageContext: 'destination-accessor-vcap'
});
/**
 * Tries to build a destination from a service binding with the given name.
 * Throws an error if no services are bound at all, no service with the given name can be found, or the service type is not supported.
 * The last error can be circumvent by using the second parameter to provide a custom function that transforms a service binding to a destination.
 * @param serviceInstanceName - The name of the service.
 * @param options - Options to customize the behavior of this function.
 * @returns A destination.
 */
async function destinationForServiceBinding(serviceInstanceName, options = {}) {
    if (options.useCache) {
        const fromCache = await destination_cache_1.destinationCache.retrieveDestinationFromCache((0, jwt_1.decodeOrMakeJwt)(options.jwt), serviceInstanceName, 'tenant');
        if (fromCache) {
            return fromCache;
        }
    }
    const selected = (0, environment_accessor_1.getServiceBindingByInstanceName)(serviceInstanceName);
    const destination = options.serviceBindingTransformFn
        ? await options.serviceBindingTransformFn(selected, options)
        : await transform(selected, options);
    const destWithProxy = destination &&
        (0, destination_service_types_1.isHttpDestination)(destination) &&
        ['internet', 'private-link'].includes((0, http_proxy_util_1.proxyStrategy)(destination))
        ? (0, http_proxy_util_1.addProxyConfigurationInternet)(destination)
        : destination;
    if (options.useCache) {
        // As the grant type is clientCredential, isolation strategy is 'tenant'.
        await destination_cache_1.destinationCache.cacheRetrievedDestination((0, jwt_1.decodeOrMakeJwt)(options.jwt), destWithProxy, 'tenant');
    }
    return destWithProxy;
}
exports.destinationForServiceBinding = destinationForServiceBinding;
async function transform(service, options) {
    if (!service_binding_to_destination_1.serviceToDestinationTransformers[service.label]) {
        throw serviceTypeNotSupportedError(service);
    }
    return service_binding_to_destination_1.serviceToDestinationTransformers[service.label](service, options);
}
function serviceTypeNotSupportedError(service) {
    return Error(`The service "${service.name}" is of type "${service.label}" which is not supported! Consider providing your own transformation function when calling destinationForServiceBinding, like this:
  destinationServiceForBinding(yourServiceName, { serviceBindingToDestination: yourTransformationFunction });`);
}
/**
 * @internal
 */
async function searchServiceBindingForDestination({ iss, jwt, serviceBindingTransformFn, destinationName, useCache }) {
    logger.debug('Attempting to retrieve destination from service binding.');
    try {
        const jwtFromOptions = iss ? { iss } : jwt ? (0, jwt_1.decodeJwt)(jwt) : undefined;
        const useCacheIgnoringUndefined = typeof useCache !== 'undefined' ? { useCache } : {};
        const destination = await destinationForServiceBinding(destinationName, {
            serviceBindingTransformFn,
            jwt: jwtFromOptions,
            ...useCacheIgnoringUndefined
        });
        logger.info('Successfully retrieved destination from service binding.');
        return destination;
    }
    catch (error) {
        logger.debug(`Could not retrieve destination from service binding. If you are not using SAP Extension Factory, this information probably does not concern you. ${error.message}`);
    }
    return null;
}
exports.searchServiceBindingForDestination = searchServiceBindingForDestination;
